src/cli/Hibp.cpp

```diff
@@ -0,0 +1,221 @@
+
+        QString path;
+        for (int i = ancestors.size() - 1; i >= 0; --i) {
+            path += ancestors[i];
+            path += " > ";
+        }
+        path += entry.title();
+
+        return QObject::tr("Password for entry '%1' has been leaked %2 times!").arg(path).arg(numPwns);
+    }
+
+} // namespace
+
+Hibp::Hibp()
+{
+    name = "hibp";
+    description = QObject::tr("Check which passwords have been leaked.");
+}
+
+int Hibp::execute(const QStringList& arguments)
+{
+    TextStream outputTextStream(Utils::STDOUT, QIODevice::WriteOnly);
+    TextStream errorTextStream(Utils::STDERR, QIODevice::WriteOnly);
+
+    QCommandLineParser parser;
+    parser.setApplicationDescription(description);
+    parser.addPositionalArgument("database", QObject::tr("Path of the password database."));
+    parser.addPositionalArgument("sha1-txt",
+                                 QObject::tr("Path of file listing SHA-1 hashes of leaked passwords, in HIBP format, "
+                                             "as available from https://haveibeenpwned.com/Passwords."));
+    parser.addOption(Command::QuietOption);
+    parser.addOption(Command::KeyFileOption);
+    parser.addHelpOption();
+    parser.process(arguments);
+
+    const QStringList args = parser.positionalArguments();
+    if (args.size() != 2) {
+        errorTextStream << parser.helpText().replace("keepassxc-cli", "keepassxc-cli hibp");
+        return EXIT_FAILURE;
+    }
+
+    const QString& dbArg = args.at(0);
+    const QString& hibpArg = args.at(1);
+
+    auto db = Utils::unlockDatabase(dbArg,
+                                    parser.value(Command::KeyFileOption),
+                                    parser.isSet(Command::QuietOption) ? Utils::DEVNULL : Utils::STDOUT,
+                                    Utils::STDERR);
+    if (!db) {
+        return EXIT_FAILURE;
+    }
+
+    QMultiHash<Sha1Hash, const Entry*> entriesBySha1;
+    for (const Entry* entry : db->rootGroup()->entriesRecursive()) {
+        const QByteArray sha1 = CryptoHash::hash(entry->password().toUtf8(), CryptoHash::Sha1);
+        entriesBySha1.insert(sha1, entry);
+    }
+
+    QFile hibpFile(hibpArg);
+    if (!hibpFile.open(QFile::ReadOnly)) {
+        errorTextStream << QObject::tr("Failed to open %1: %2").arg(hibpArg).arg(hibpFile.errorString()) << endl;
+        return EXIT_FAILURE;
+    }
+
+    quint64 lineNum = 1;
+    while (true) {
+        Sha1Hash sha1;
+        size_t numPwns = 0;
+        bool atEnd = false;
+        QString error;
+        if (!parseHibpLine(hibpFile, sha1, numPwns, atEnd, error)) {
+            errorTextStream << QObject::tr("HIBP file, line %1: %2").arg(lineNum).arg(error) << endl;
+            return EXIT_FAILURE;
+        }
+        if (atEnd) {
+            break;
+        }
+
+        for (auto it = entriesBySha1.find(sha1); it != entriesBySha1.end() && it.key() == sha1; ++it) {
+            outputTextStream << pwnageReport(*it.value(), numPwns) << endl;
```

Comment:
Do you really expect to find multiple entries with the same SHA-1 hash?
Also, shouldn't you output only pwned entries? This list could become awfully long.

Reply:
Whenever two entries have the same password, they'll have the same SHA-1 of course, and I want to output _all_ matching entries here.

And I am only emitting pwned entries -- the `sha1` I'm searching by comes from the HIBP file over whose lines we're iterating over.

Reply:
True, ignore this comment.